File: src/test/java/toby/command/commands/music/SetVolumeCommandTest.java
package toby.command.commands.music;

import net.dv8tion.jda.api.entities.emoji.RichCustomEmoji;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import toby.command.CommandContext;
import toby.emote.Emotes;

import static org.mockito.Mockito.*;

class SetVolumeCommandTest implements MusicCommandTest {

    SetVolumeCommand setVolumeCommand;

    @BeforeEach
    void setUp() {
        setupCommonMusicMocks();
        setVolumeCommand = new SetVolumeCommand();
    }

    @AfterEach
    void tearDown() {
        tearDownCommonMusicMocks();
    }

    @Test
    void testSetVolume_withValidArgs() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 20;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(true);
        int oldVolume = 21;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(1)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessageFormat(eq("Changing volume from '%s' to '%s' \uD83D\uDD0A"), eq(oldVolume), eq(volumeArg));
    }

    @Test
    void testSetVolume_withOldAndNewVolumeBeingTheSame_SendsErrorMessage() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 20;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(true);
        int oldVolume = 20;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(0)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessageFormat(eq("New volume and old volume are the same value, somebody shoot %s"), eq("Effective Name"));
    }

    @Test
    void testSetVolume_withNewVolumeBeingOver100_SendsErrorMessage() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 101;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(true);
        int oldVolume = 20;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(0)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessage(eq("Set the volume of the audio player for the server to a percent value (between 1 and 100)"));
    }

    @Test
    void testSetVolume_withNewVolumeBeingNegative_SendsErrorMessage() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 101;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(true);
        int oldVolume = 20;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(0)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessage(eq("Set the volume of the audio player for the server to a percent value (between 1 and 100)"));
    }

    @Test
    void testSetVolume_withInvalidPermissions() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 20;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(true);
        int oldVolume = 21;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);
        when(requestingUserDto.hasMusicPermission()).thenReturn(false);
        RichCustomEmoji tobyEmote = mock(RichCustomEmoji.class);
        when(jda.getEmojiById(Emotes.TOBY)).thenReturn(tobyEmote);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(0)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessageFormat(eq("You aren't allowed to change the volume kid %s"), eq(tobyEmote));
    }

    @Test
    void testSetVolume_whenSongIsNotStoppableAndWithoutOverridingPermissions_SendsError() {
        //Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        OptionMapping volumeOptionMapping = mock(OptionMapping.class);
        when(event.getOption("volume")).thenReturn(volumeOptionMapping);
        int volumeArg = 20;
        when(volumeOptionMapping.getAsInt()).thenReturn(volumeArg);
        when(playerManager.isCurrentlyStoppable()).thenReturn(false);
        int oldVolume = 21;
        when(audioPlayer.getVolume()).thenReturn(oldVolume);
        when(requestingUserDto.isSuperUser()).thenReturn(false);
        RichCustomEmoji tobyEmote = mock(RichCustomEmoji.class);
        when(jda.getEmojiById(Emotes.TOBY)).thenReturn(tobyEmote);

        //Act
        setVolumeCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(audioPlayer, times(0)).setVolume(volumeArg);
        verify(interactionHook, times(1)).sendMessageFormat(eq("You aren't allowed to change the volume kid %s"), eq(tobyEmote));
    }
}

File: src/test/kotlin/toby/command/commands/music/SetVolumeCommandTest.kt
package toby.command.commands.music

import io.mockk.*
import net.dv8tion.jda.api.entities.emoji.RichCustomEmoji
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import toby.command.CommandContext
import toby.command.CommandTest
import toby.command.CommandTest.Companion.event
import toby.emote.Emotes

internal class SetVolumeCommandTest : MusicCommandTest {
    lateinit var setVolumeCommand: SetVolumeCommand

    @BeforeEach
    fun setUp() {
        setupCommonMusicMocks()
        setVolumeCommand = SetVolumeCommand()
    }

    @AfterEach
    fun tearDown() {
        tearDownCommonMusicMocks()
        clearAllMocks()
    }

    @Test
    fun testSetVolume_withValidArgs() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = 20
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns true
        val oldVolume = 21
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 1) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            event.hook.sendMessageFormat(
                "Changing volume from '%s' to '%s' \uD83D\uDD0A",
                oldVolume,
                volumeArg
            )
        }
    }

    @Test
    fun testSetVolume_withOldAndNewVolumeBeingTheSame_SendsErrorMessage() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = 20
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns true
        val oldVolume = 20
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 0) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            event.hook.sendMessageFormat(
                "New volume and old volume are the same value, somebody shoot %s",
                "Effective Name"
            )
        }
    }

    @Test
    fun testSetVolume_withNewVolumeBeingOver100_SendsErrorMessage() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = 101
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns true
        val oldVolume = 20
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 0) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            event.hook.sendMessage(
                "Set the volume of the audio player for the server to a percent value (between 1 and 100)"
            )
        }
    }

    @Test
    fun testSetVolume_withNewVolumeBeingNegative_SendsErrorMessage() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = -1
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns true
        val oldVolume = 20
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 0) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            event.hook.sendMessage(
                "Set the volume of the audio player for the server to a percent value (between 1 and 100)"
            )
        }
    }

    @Test
    fun testSetVolume_withInvalidPermissions() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = 20
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns true
        val oldVolume = 21
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume
        every { CommandTest.requestingUserDto.musicPermission } returns false
        val tobyEmote = mockk<RichCustomEmoji>()
        every { CommandTest.jda.getEmojiById(Emotes.TOBY) } returns tobyEmote

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 0) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            event.hook.sendMessageFormat(
                "You aren't allowed to change the volume kid %s",
                tobyEmote
            )
        }
    }

    @Test
    fun testSetVolume_whenSongIsNotStoppableAndWithoutOverridingPermissions_SendsError() {
        // Arrange
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        val volumeOptionMapping = mockk<OptionMapping>()
        every { event.getOption("volume") } returns volumeOptionMapping
        val volumeArg = 20
        every { volumeOptionMapping.asInt } returns volumeArg
        every { MusicCommandTest.playerManager.isCurrentlyStoppable } returns false
        val oldVolume = 21
        every { MusicCommandTest.audioPlayer.volume } returns oldVolume
        every { CommandTest.requestingUserDto.superUser } returns false
        val tobyEmote = mockk<RichCustomEmoji>()
        every { CommandTest.jda.getEmojiById(Emotes.TOBY) } returns tobyEmote
        val hook = event.hook

        // Act
        setVolumeCommand.handleMusicCommand(
            commandContext,
            MusicCommandTest.playerManager,
            CommandTest.requestingUserDto,
            0
        )

        // Assert
        verify(exactly = 0) { MusicCommandTest.audioPlayer.volume = volumeArg }
        verify(exactly = 1) {
            hook.sendMessageFormat(
                "You aren't allowed to change the volume kid %s",
                tobyEmote
            )
        }
    }
}
